Applications/SpaceAnalyzer/main.cpp

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "TreeMapWidget.h"
+#include <AK/Queue.h>
+#include <AK/QuickSort.h>
+#include <AK/RefCounted.h>
+#include <Applications/SpaceAnalyzer/SpaceAnalyzerGML.h>
+#include <LibCore/DirIterator.h>
+#include <LibCore/File.h>
+#include <LibGUI/AboutDialog.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/BreadcrumbBar.h>
+#include <LibGUI/Icon.h>
+#include <LibGUI/Menu.h>
+#include <LibGUI/MenuBar.h>
+#include <LibGUI/StatusBar.h>
+#include <sys/stat.h>
+
+static const char* APP_NAME = "SpaceAnalyzer";
+
+struct TreeNode : public TreeMap::ITreeMapNode {
+    TreeNode(String name)
+        : m_name(name)
+        , m_area(0)
+        , m_children(nullptr) {};
+    String m_name;
+    int64_t m_area;
+    OwnPtr<Vector<TreeNode>> m_children;
```

Comment:
Is there a meaningful reason to allocate these vectors on the heap instead of making it simply `Vector<TreeNode> m_children;`?

Reply:
I tried this after your suggestion, but it doesn't work because Vector<T> does sizeof(T) here: https://github.com/SerenityOS/serenity/blob/ddc255a6c0026af70a7ca685c2fed24941e7df9c/AK/Vector.h#L608
But T is an incomplete type at that point.

My original reasoning however, was to reduce the total size of the tree. I guessed (did not measure) that most nodes will be leaf nodes. Having a nullptr in those nodes would be smaller than an empty Vector.

Reply:
Oh duh, my bad, I totally missed that it's a vector of TreeNode *inside* TreeNode :)